chisel/src/config.rs

```diff
@@ -0,0 +1,553 @@
+                        Some(option.to_string())
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            // Prepare the configuration map for this specific module.
+            let mut opts_final = HashMap::new();
+
+            // Now we have all the relevant options, still formatted as "module.option=value"
+            for opt in relevant_options {
+                // Get a string for all chars after the first dot, which should yield
+                // "option=value"
+                let keyval_pair: String = opt.chars().skip_while(|c| *c != '.').skip(1).collect();
+
+                // Split into its terms.
+                let keyval_pair: Vec<&str> = keyval_pair.split('=').collect();
+
+                if keyval_pair.len() == 2 {
+                    chisel_debug!(
+                        1,
+                        "Found option for module {}: {} = {}",
+                        name,
+                        &keyval_pair[0],
+                        &keyval_pair[1]
+                    );
+                    opts_final.insert(keyval_pair[0].to_string(), keyval_pair[1].to_string());
+                } else {
+                    return Err(format!(
+                        "In module {}: option {} malformed: unexpected equals sign",
+                        name, keyval_pair[0]
+                    ));
+                }
+            }
+
+            // Add the complete module config.
+            module_configs.push_back((name.to_string(), ModuleConfig(opts_final)));
+        }
+
+        // Since this is used in oneliner mode, the only useful "top-level" option is the input
+        // file. We set this manually in oneliner mode, so we can leave this empty.
+        let opts: HashMap<String, String> = HashMap::new();
+        Ok(Ruleset {
+            opts,
+            modules: module_configs,
+        })
+    }
+}
+
+impl FromYaml for ChiselConfig {
+    fn from_yaml(yaml: &Value) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let rulesets = yaml.as_mapping().expect("Already validated");
+        let rulesets: VecDeque<(String, Ruleset)> = rulesets
+            .iter()
+            .map(|(key, val)| {
+                (
+                    key.as_str().expect("Already validated").to_string(),
+                    Ruleset::from_yaml(val).expect("No failure cases yet"),
+                )
+            })
+            .collect();
+
+        Ok(ChiselConfig(rulesets))
+    }
+}
+
+impl FromYaml for Ruleset {
+    fn from_yaml(yaml: &Value) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let mut opts: HashMap<String, String> = HashMap::new();
+        let mut modules: VecDeque<(String, ModuleConfig)> = VecDeque::new();
+
+        let ruleset = yaml.as_mapping().expect("Already validated");
+        for (key, val) in ruleset.iter() {
+            match (key, val) {
```

Comment:
I think this match currently is not adding any benefit, unless the `String, any` case can be broken into `String, mapping` and `String, null` - otherwise we mix two stiles of programming to make it harder to read.

Reply:
But fine with it this way too. Not sure what's better.

Reply:
I think there isn't any specific logic which would warrant duplicating the code into two cases.